sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -480,10 +480,120 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
       case expressions.Contains(a: Attribute, Literal(v: UTF8String, StringType)) =>
         Some(sources.StringContains(a.name, v.toString))
 
+      case expressions.BinaryComparison(BinaryArithmetic(left, right), Literal(v, t)) =>
+        translateArithemiticOPFilter (predicate)
+      case expressions.BinaryComparison(Literal(v, t), BinaryArithmetic(left, right)) =>
+        translateArithemiticOPFilter (predicate)
+        
       case _ => None
     }
   }
 
+  private def translateArithemiticOPFilter(predicate: Expression): Option[Filter] = {
+    predicate match {
+      case expressions.EqualTo(Add(left, right), Literal(v, t)) =>
+        Some(sources.ArithmeticOPEqualTo(Add(left, right), convertToScala(v, t)))
```

Comment:
As described in [SPARK-10195](https://issues.apache.org/jira/browse/SPARK-10195), it looks now data sources API exposes Catalyst's internal types through its Filter interfaces. I think this might have to be hidden.